Enable unwinding on ARM

yuyichao opened the pull request:
There are multiple issues that hold this back, most of them from upstream.
1. GCC 5.0+ miscompiles LLVM on ARM (i.e. https://github.com/JuliaLang/julia/issues/14550)
   
   Now that I've found a workaround, I can finally start playing with unwinding on ARM.
   (In another word, users of LLVM compiled with GCC 5.0+ without LTO and with optimization will break.)
2. ARM use a special format other than DWARF for unwind info.
   
   The first commit makes sure the unwind info for JIT code is registered with libunwind. This is the only change we need to make to our code.
3. LLVM emit incorrect unwind info (wrong relocation).
   
   Fixed by the LLVM patch. Submitted upstream at https://reviews.llvm.org/D25069. Should be the right approach but needs to be approved and I probably need to figure out how to write a test for it....
4. libunwind doesn't accept dynamic unwind info registration on ARM, messes up unwind info lookup of different formats, and doesn't like DWARF 4 debug info....
   
   The first patch is submitted to the libunwind-devel list, the second and the third are a tiny bit more hacky and we'll probably not wait for upstream to accept this anyway.....

Local test shows that the unwinding is pretty reliable now and the tests this fixes are `backtrace`, `replutil`, `meta`, `stacktraces`, `cmdlineargs`, `compile`. The ones that are still failing are `profile` and `parallel`. I'll check why those are still failing later but the PR should be good to go without that.

yuyichao committed:
Enable unwinding on ARM

Implement dynamic unwind info registration on ARM

yuyichao committed:
Add LLVM patch to fix unwind info relocation

yuyichao committed:
Add libunwind patches to enable and fix dynamic unwind info registration on ARM

yuyichao committed:
Fix signal context to libunwind context conversion on ARM

tkelman commented on deps/unwind.mk:
```diff
@@ -12,7 +12,14 @@ $(SRCDIR)/srccache/libunwind-$(UNWIND_VER)/source-extracted: $(SRCDIR)/srccache/
 	echo 1 > $@
 
-$(BUILDDIR)/libunwind-$(UNWIND_VER)/build-configured: $(SRCDIR)/srccache/libunwind-$(UNWIND_VER)/source-extracted
+$(SRCDIR)/srccache/libunwind-$(UNWIND_VER)/libunwind-patch-applied: $(SRCDIR)/srccache/libunwind-$(UNWIND_VER)/source-extracted
```
we're probably going to add more of these, best to track individually

yuyichao replied:
For tracking dependencies and auto recompilation?

tkelman replied:
to handle patches that get added later

ViralBShah commented:
Is this possible to backport to 0.5?

yuyichao commented:
The tests can all pass now. Profiling is not really reliable though (in that the test can pass but I get segfault ~50% of the time on the first run). This is possibly due to the use of non-signal-safe functions in libunwind.

Not sure about backporting. The only regressions I can think of are the GCC 5.0+ but and the `@profile` could segfault instead of doing nothing.....

yuyichao commented:
AFAICT the travis failure is due to the change in the timestamp file name compare to the previous commit, is there a way to clear the libunwind cache without clearing others? (push a dummy commit that clears it?)

vtjnash commented:
I don't see a PR-relative cache file on travis, maybe just try restarting it?

yuyichao commented:
Although I guess it shouldn't be too bad if I just clear the PR cache? This only touches LLVM and Libunwind so I don't need to rebuild all dependencies.

yuyichao commented:
> I don't see a PR-relative cache file on travis, maybe just try restarting it?

Hmmm, you are right. Wierd.

yuyichao commented:
Restarted. Old log here https://gist.github.com/yuyichao/7ddfe4f82472ddf2d313e067b3341443

yuyichao commented:
It seems to be using PR cache according to the log though?

```
5.98sattempting to download cache archive

fetching PR.18761/cache-linux-precise-d1f8e8c389023adf2674031c40577df6c1e5492bd1f9ff85c6adfdf7268c67e4--compiler-gpp-5-m32.tgz

fetching PR.18761/cache--compiler-gpp-5-m32.tgz

fetching master/cache-linux-precise-d1f8e8c389023adf2674031c40577df6c1e5492bd1f9ff85c6adfdf7268c67e4--compiler-gpp-5-m32.tgz

found cache
```

yuyichao commented:
I added a LLVM test and rebased. Should we wait for the LLVM patch to be merged first? The review request didn't get any comment from the maintainers over the past week.

ViralBShah commented:
We can merge here since we have a patch file. Do we really need to wait for upstream to merge?

yuyichao commented:
The LLVM PR has passed review and I'll try to get someone ( @Keno ) to push it for me soon. Will merge this PR tomorrow.
